centrallix/expression/exp_functions.c

```diff
@@ -4159,6 +4161,92 @@ int exp_fn_cos_compare(pExpression tree, pParamObjects objlist, pExpression i0,
     return 0;
     }
 
+
+/*
+ * exp_fn_argon2id
+ * This method hashes a given password using the Argon2 algorithm (ID variant)
+ *
+ * Parameters:
+ * 	pExpression tree: 
+ * 	pParamObjects: 
+ * 	pExpression passowrd: The password, passed as a pExpression
+ * 	pExpression salt: The salt, passed as a pExpression
+ *
+ * returns:
+ *	 0 if successful
+ *	 -1 if error
+ */
+int exp_fn_argon2id(pExpression tree, pParamObjects objlist, pExpression password, pExpression salt)
+{
+    // Check for and set the optional parameters
+    // T_COST controls the time cost
+    // M_COST controls the memory cost
+    // PARALLELISM controls how many 'lanes' will be used
+    // HASHLEN is the size of the finished hash
+    unsigned int T_COST = (tree->Children.nItems >= 3)?((intptr_t)tree->Children.Items[2]):2;      // the default value should be changed according to system needs
+    unsigned int M_COST = (tree->Children.nItems >= 4)?((intptr_t)tree->Children.Items[3]):(1<<16);      // the default value should be changed according to system needs
+    unsigned int PARALLELISM = (tree->Children.nItems >= 5)?((intptr_t)tree->Children.Items[4]):1; // the default value should be changed according to system needs
+    unsigned int HASHLEN = (tree->Children.nItems >= 6)?((intptr_t)tree->Children.Items[5]):32;    // the default value should be changed according to system needs
+    
```

Comment:
To prevent a denial of service condition, we should make sure there are bounds on acceptable values for these four parameters.  Lower bounds are whatever makes sense for valid values for argon2id.  Upper bounds:  maybe 1 second for time, 32MB for RAM, 8 threads for parallelism, and 256 bytes for hashlen?  We can always fix those later if those limits aren't appropriate.

For a later project, it might be good to only allow a user to have one argon() running at a time system-wide, to prevent a user from bogging down / DoSing the server with argon() calls since it is intentionally an expensive function.  But that would require adding a mutex / semaphore to a session somewhere, and I'd have to track down where'd be the best place to put that.

Reply:
Sounds good. That's easy to add.

Reply:
I should add -- the above code doesn't look correct either.  The integer would be ((pExpression)tree->Children.Items[3])->Integer, but you'd also need to verify the datatype is DATA_T_INTEGER and the flags don't contain EXPR_F_NULL.

Reply:
Okay. I'll make those changes as well.